rust/agama-network/src/dbus/tree.rs

```diff
@@ -0,0 +1,206 @@
+use agama_lib::error::ServiceError;
+use parking_lot::Mutex;
+use uuid::Uuid;
+
+use crate::{action::Action, dbus::interfaces, model::*};
+use std::collections::HashMap;
+use std::sync::mpsc::Sender;
+use std::sync::Arc;
+
+const CONNECTIONS_PATH: &str = "/org/opensuse/Agama/Network1/connections";
+const DEVICES_PATH: &str = "/org/opensuse/Agama/Network1/devices";
+
+/// Handle the objects in the D-Bus tree for the network state
+pub struct Tree {
+    connection: zbus::Connection,
+    actions: Sender<Action>,
+    objects: Arc<Mutex<ObjectsRegistry>>,
+}
+
+impl Tree {
+    /// Creates a new tree handler.
+    ///
+    /// * `connection`: D-Bus connection to use.
+    /// * `actions`: sending-half of a channel to send actions.
+    pub fn new(connection: zbus::Connection, actions: Sender<Action>) -> Self {
+        Self {
+            connection,
+            actions,
+            objects: Default::default(),
+        }
+    }
+
+    /// Adds devices to the D-Bus tree.
+    ///
+    /// * `devices`: list of devices.
+    pub async fn add_devices(&mut self, devices: &Vec<Device>) -> Result<(), ServiceError> {
+        for (i, dev) in devices.iter().enumerate() {
```

Comment:
this is kind of strange or I do not get logic, but if I call this method twice, then result is it that later ones rewrite ones from first call?